File: src/Cpp/15_Funkcje_rekurencja/Zad9.cpp
#include <algorithm>
#include <cassert>
#include <string>

/*
Otrzymuejsz napi. Sprawdź przy pomocy rekurencji czy otrzymane słowo 
jest słowem elfickim. Przez słowo elfickie rozumiemy taki napis, w 
którym co najmniej raz występuje każda z liter słowa elf.
*/


int znajdz(std::string slowo, char znak, unsigned int pozycja=0)
{
   /**
   * Funkcja zwraca indeks pierwszego wystapienia litera w slowie.
   */
    if (pozycja >= slowo.size())
      return -1;
    
    if (slowo[pozycja] == znak)
      return pozycja;
    
    return znajdz(slowo, znak, pozycja + 1);
}


bool czySlowoElfickie(std::string slowo, std::string elf = "elf") {
  /**
   * Funkcja sprawdza czy wszystkie litery slowa elf wystepuja w napisie.
   */
  
  if (elf.empty())
    return true;
  
  if (slowo.empty())
    return false;
  
  auto pozycja = znajdz(slowo, elf[0]);

  if (pozycja == -1)
    return false;

  // usun znak na pozycji pozycja z napisu slowo
  slowo.erase(pozycja, 1);

  return czySlowoElfickie(slowo, elf.substr(1));
}

void test1() {
  std::string slowo = "reflektor";
  bool wynik = true;

  assert(czySlowoElfickie(slowo) == wynik);
}

void test2() {
  std::string slowo = "elzbieta";
  bool wynik = false;

  assert(czySlowoElfickie(slowo) == wynik);
}

int main() {

  test1();
  test2();

  return 0;
}

File: src/Python/15_Funkcje_rekurencja/Zad9.py
"""
Otrzymuejsz napis. Sprawdz przy pomocy rekurencji czy otrzymane slowo 
jest slowem elfickim. Przez slowo elfickie rozumiemy taki napis, w 
ktorym co najmniej raz wystepuje kazda z liter slowa elf.
"""


def znajdz(slowo, litera, pozycja=0):
    """
    Funkcja zwraca indeks pierwszego wystapienia litera w slowie.
    """
    if len(slowo) <= pozycja:
        return -1
    if slowo[pozycja] == litera:
        return pozycja
    return znajdz(slowo, litera, pozycja + 1)


def czy_slowo_elfickie(napis, elf="elf"):
    """
    Funkcja sprawdza czy wszystkie litery slowa elf wystepuja w napisie.
    """

    if len(elf) == 0:
        return True

    if len(napis) == 0:
        return False

    pozycja = znajdz(napis, elf[0])

    if pozycja == -1:
        return False

    return czy_slowo_elfickie(napis[:pozycja] + napis[pozycja + 1 :], elf[1:])


def test_czy_slowo_elfickie():
    assert czy_slowo_elfickie("elf")
    assert czy_slowo_elfickie("reflektor")
    assert not czy_slowo_elfickie("elzbieta")


if __name__ == "__main__":

    test_czy_slowo_elfickie()
